core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -231,31 +227,24 @@ private object RTS {
       }
     }
 
-    private final def collectDefect[E, A](e: ExitResult[E, A]): Option[List[Throwable]] =
-      e match {
-        case ExitResult.Terminated(ts @ _ :: _) => Some(ts)
-        case ExitResult.Failed(_, ts @ _ :: _)  => Some(ts)
-        case _                                  => None
-      }
-
     /**
-     * Catches an exception, returning a (possibly null) finalizer action that
-     * must be executed. It is painstakingly *guaranteed* that the stack will be
-     * empty in the sole case the exception was not caught by any exception
-     * handler—i.e. the exceptional case.
+     * Empties the stack, collecting all finalizers and coalescing them into an
+     * action that produces a list (possibly empty) of errors during finalization.
+     * If needed, catch exceptions and apply redeem error handling.
      */
-    final def catchError: IO[Nothing, Option[List[Throwable]]] = {
+    final def interruptStack(catchError: Boolean): IO[Nothing, Option[List[Throwable]]] = {
       var errorHandler: Any => IO[Any, Any]               = null
       var finalizer: IO[Nothing, Option[List[Throwable]]] = null
 
       // Unwind the stack, looking for exception handlers and coalescing
       // finalizers.
       while ((errorHandler eq null) && !stack.isEmpty) {
         stack.pop() match {
-          case a: IO.Redeem[_, _, _, _] =>
+          case a: IO.Redeem[_, _, _, _] if catchError =>
```

Comment:
I like how, at long last, `catchError` can be unified with `interruptStack`! 😍 

They had basically the same logic.